components/movementsensor/gpsrtk/pmtk.go

```diff
@@ -0,0 +1,197 @@
+	}
+	return nil
+}
+
+// validateNtrip ensures all parts of the config are valid.
+func (cfg *I2CConfig) validateNtrip(path string) error {
+	if cfg.NtripURL == "" {
+		return resource.NewConfigValidationFieldRequiredError(path, "ntrip_url")
+	}
+	return nil
+}
+
+func init() {
+	resource.RegisterComponent(
+		movementsensor.API,
+		i2cRtkmodel,
+		resource.Registration[movementsensor.MovementSensor, *I2CConfig]{
+			Constructor: newRTKI2C,
+		})
+}
+
+func newRTKI2C(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (movementsensor.MovementSensor, error) {
+	return makeRTKI2C(ctx, deps, conf, logger, nil)
+}
+
+// makeRTKI2C is separate from newRTKI2C, above, so we can pass in a non-nil mock I2C bus during
+// unit tests.
+func makeRTKI2C(
+	ctx context.Context,
+	_ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+	mockI2c buses.I2C,
+) (movementsensor.MovementSensor, error) {
+	newConf, err := resource.NativeConfig[*I2CConfig](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	g := &gpsrtk{
+		Named:      conf.ResourceName().AsNamed(),
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+		err:        movementsensor.NewLastError(1, 1),
+	}
+
+	ntripConfig := &gpsutils.NtripConfig{
+		NtripURL:             newConf.NtripURL,
+		NtripUser:            newConf.NtripUser,
+		NtripPass:            newConf.NtripPass,
+		NtripMountpoint:      newConf.NtripMountpoint,
+		NtripConnectAttempts: newConf.NtripConnectAttempts,
+	}
+
+	g.ntripClient, err = gpsutils.NewNtripInfo(ntripConfig, g.logger)
+	if err != nil {
+		return nil, err
+	}
+	g.InputProtocol = "i2c"
+
+	config := gpsutils.I2CConfig{
+		I2CBus:      newConf.I2CBus,
+		I2CBaudRate: newConf.I2CBaudRate,
+		I2CAddr:     newConf.I2CAddr,
+	}
+	if config.I2CBaudRate == 0 {
+		config.I2CBaudRate = 115200
+	}
+
+	// If we have a mock I2C bus, pass that in, too. If we don't, it'll be nil and constructing the
+	// reader will create a real I2C bus instead.
+	dev, err := gpsutils.NewI2cDataReader(config, mockI2c, logger)
+	if err != nil {
+		return nil, err
+	}
+	g.cachedData = gpsutils.NewCachedData(dev, logger)
+
+	g.correctionWriter, err = newCorrectionWriter(newConf.I2CBus, byte(newConf.I2CAddr))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := g.start(); err != nil {
+		return nil, err
+	}
+
+	// It's possible that we've taken so long to start up that the resource manager has given up on
+	// us and tried constructing a new component instead. If that happens, we don't want 2
+	// components talking to the same chip. So, if our context is canceled, close our component
+	// instead of returning it.
+	if ctx.Err() != nil {
+		logger.Warn("context canceled by the end of the constructor! Closing the new component...")
+		return nil, fmt.Errorf("timed out constructing I2C RTK reader. Closing: %w", g.Close(ctx))
+	}
```

Comment:
This is... weird; I would want the resource manager to be the arbitrator of cancelling the resource node's creation and ensuring its closure rather than the resource closing itself upon a context-cancelled error in the creator, but if you're seeing this behaviour, we should guard against it. Maybe make a ticket for netcode to handle closures better upon context cancellation in resource creation? This sounds like a bug rather than something we expect users and driver implementers to do.

Reply:
Done: https://viam.atlassian.net/browse/RSDK-8146

When they address that ticket, hopefully we can remove this paragraph. but it's here in the meantime.